pkg/apiserver/rest/dualwriter_mode3.go

```diff
@@ -42,30 +44,61 @@ func (d *DualWriterMode3) Mode() DualWriterMode {
 const mode3Str = "3"
 
 // Create overrides the behavior of the generic DualWriter and writes to LegacyStorage and Storage.
-func (d *DualWriterMode3) Create(ctx context.Context, obj runtime.Object, createValidation rest.ValidateObjectFunc, options *metav1.CreateOptions) (runtime.Object, error) {
+func (d *DualWriterMode3) Create(ctx context.Context, in runtime.Object, createValidation rest.ValidateObjectFunc, options *metav1.CreateOptions) (runtime.Object, error) {
 	var method = "create"
 	log := d.Log.WithValues("method", method)
+
 	ctx = klog.NewContext(ctx, log)
 
-	startStorage := time.Now()
-	created, err := d.Storage.Create(ctx, obj, createValidation, options)
+	accIn, err := meta.Accessor(in)
 	if err != nil {
+		return nil, err
+	}
+
+	if accIn.GetUID() != "" {
+		return nil, fmt.Errorf("UID should not be: %v", accIn.GetUID())
+	}
+
+	if accIn.GetName() == "" && accIn.GetGenerateName() == "" {
+		return nil, fmt.Errorf("name or generatename have to be set")
+	}
+
+	startStorage := time.Now()
+	storageObj, errObjectSt := d.Storage.Create(ctx, in, createValidation, options)
+	d.recordStorageDuration(errObjectSt != nil, mode3Str, d.resource, method, startStorage)
+	if errObjectSt != nil {
 		log.Error(err, "unable to create object in storage")
-		d.recordLegacyDuration(true, mode3Str, d.resource, method, startStorage)
-		return created, err
+		return storageObj, errObjectSt
 	}
-	d.recordStorageDuration(false, mode3Str, d.resource, method, startStorage)
 
-	go func() {
-		ctx, cancel := context.WithTimeoutCause(ctx, time.Second*10, errors.New("legacy create timeout"))
-		defer cancel()
+	createdCopy := storageObj.DeepCopyObject()
 
-		startLegacy := time.Now()
-		_, errObjectSt := d.Legacy.Create(ctx, obj, createValidation, options)
-		d.recordLegacyDuration(errObjectSt != nil, mode3Str, d.resource, method, startLegacy)
-	}()
+	//nolint:errcheck
+	go d.createOnLegacyStorage(ctx, in, createdCopy, createValidation, options)
```

Comment:
This should not be best effort in a goroutine. IMO in mode 3 we should ensure the writes to legacy are always successful. Basically, CRUD are the same in Mode 2 and mode 3 ? Only GET/LIST are different. WDYT ?

Reply:
I agree, otherwise we can't roll back from mode 3 to 2 safely

Reply:
we can do that, but it's really out of the scope of this PR, as we are now referring of how DW modes should behave. At some point it was agreed we could have a symmetrical behaviour: mode 1 as best effort to US, mode 2 fails on either and mode 3 is best effort to legacy. We can change that as we continue to improve of how each mode should behave, but let's do it in another PR :)